hashicorp/consul
# Backport of xds: don't attempt to load-balance sessions for local proxies into release/1.14.x

## Backport

This PR is auto-generated from #15789 to be assessed for backporting due to the inclusion of the label backport/1.14.


**WARNING** automatic cherry-pick of commits failed. Commits will require human attention.

> merge conflict error: POST https://api.github.com/repos/hashicorp/consul/merges: 409 Merge conflict []



The below text is copied from the body of the original PR.

---

### Description
Previously, we'd begin a session with the xDS concurrency limiter regardless of whether the proxy was registered in the catalog or in the server's local agent state.

This caused problems for users who run `consul connect envoy` directly against a server rather than a client agent, as the server's locally registered proxies wouldn't be included in the limiter's capacity.

Now, the `ConfigSource` is responsible for beginning the session and we only do so for services in the catalog (bypassing the limiter for locally-registered proxies).

Fixes: https://github.com/hashicorp/consul/issues/15753

Also upgrades mockery to v2.15.0, as this was required to regenerate the mocks on Go 1.19.

### Testing & Reproduction steps
* Run a Consul cluster with >1 server.
* Register the majority of the proxies to 1 server.
* Run `consul connect envoy` for each proxy.
* Previously you'd see "this server has too many xDS streams open" in the Envoy logs.
* Now it should work as expected.


---

<details>
<summary> Overview of commits </summary>

  - f269f907729a1857b923caf10bb06b44d1d0653c  - 43c91e2897548df636013deff778179cf6392f9d  - f15b4a4459114368665c5d9f4374444abd1df7ab  - 668a4d1671968bb45e72140202c012136d8878bc  - d4add27e17c9f69f8853cfc1b45f70b069483b8c  - 087db9637fbe8a03e20fdad1cdca2faa8d8f7814  - d19960cfeb7cd81e48d123d679dce41cb69e8745 

</details>
